### DISCUSSION TITLE
Is an implementation of geometrical integrators like RK4 for molecular dynamics possible in the HOOMD framework?

### DISCUSSION THREAD

**Discussion by EliasPhysics** [glotzerlab/hoomd-blue#1912] [Support] [answered]
Hello HOOMD users,

I am currently using HOOMD to simulate a swarm-like system of particles driven by a predator particle on a chaotic trajectory. The system has dissipative forces and is thus non-hamiltonian.
The problem is that I get numerical artifacts for the constant volume integration scheme for certain parameters where particle dynamics are no longer physically meaningful. I wanted to ask if there are integrators implemented in HOOMD that are more suitable for this system. I thought about using geometrical integrators like the Runge-Kutta integrator to have stable local trajectories, would that be possible to be implemented in the HOOMD framework?

Thank you in advance,
Elias

**Comment by joaander** [comment, marked as answer]
Yes, you could write a custom component that implements Runge-Kutta or any integrator you like. HOOMD-blue defines an `Integrator` as an operation that evolves the state of the simulation at some step t to step t+1: 
* https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/hoomd/Integrator.h
* https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/ARCHITECTURE.md

You can implement custom integrators in a C++ component: https://github.com/glotzerlab/hoomd-component-template/
